### PR TITLE
#827 finishing touches to group members list

### LINKED ISSUE #827: Add finishing touches to GroupMembers.vue
### Problem

Screen Shot 2020-02-08 at 3 13 14 PM

The screenshot above shows what the group members list currently looks like with 13 members. Note that some things are missing from the Figma design:

- `displayName` isn't shown for users, only `username`
- The user that's you doesn't have `(you)` next to their name
- New users do not have the `new` tag next to their name (and it's also not clear to me whether they should be at the top or not, that's something @mmbotelho can comment on)

### Solution

Address each of these issues in turn, either as multiple PRs or one big PR.

### PR DESCRIPTION
**Pull request by sandrina-p**
Closes #827 

image

- Members show `displayName` if available.
- Show "(you)" when the member is us.
- Add pill "new" next to members that joined after me. The pill is visible for **1 week**.
- Include "pending members" that were approved to be invited with the pill "pending".
- Update Cypress tests to coverage
- Add "pending members" to modal "all members".
- Note: #869 is not done in this PR. It will be another one.

#### Scouting fixes:
- Improve button loading state when reduced motion is activated.
- Remove unused .pill variants and review its margins where used.

### CODE REVIEW

**Review by mmbotelho — DISMISSED** [review]
Perfect @sandrina-p thank you!

File: frontend/model/state.js [hunk 1]
```diff
@@ -334,7 +334,7 @@ const getters = {
     return username => {
       const groupProfile = getters.groupProfile(username)
       const identityState = groupProfile && state[groupProfile.contractID]
-      return identityState && identityState.attributes
+      return identityState ? identityState.attributes : {}
```
Comment (taoeffect) [inline frontend/model/state.js]: Why was this added?

File: frontend/controller/actions/group.js [hunk 2]
```diff
@@ -98,7 +99,7 @@ export default sbp('sbp/selectors/register', {
     try {
       // post acceptance event to the group contract
       const acceptance = await sbp('gi.contracts/group/inviteAccept/create',
-        { inviteSecret },
+        { inviteSecret, joined_ms: Date.now() },
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: Remove (see other comments)

File: frontend/controller/actions/group.js [hunk 3]
```diff
@@ -48,6 +48,7 @@ export default sbp('sbp/selectors/register', {
         invites: {
           [initialInvite.inviteSecret]: initialInvite
         },
+        joined_ms: Date.now(),
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: Remove (see other comments)

File: frontend/model/contracts/group.js [hunk 4]
```diff
@@ -423,7 +437,8 @@ DefineContract({
     },
     'gi.contracts/group/inviteAccept': {
       validate: objectOf({
-        inviteSecret: string // NOTE: simulate the OP_KEY_* stuff for now
+        inviteSecret: string, // NOTE: simulate the OP_KEY_* stuff for now,
+        joined_ms: number
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: Remove (see other comments)

File: frontend/model/contracts/group.js [hunk 5]
```diff
@@ -199,7 +213,7 @@ DefineContract({
             groupCreator: meta.username
           },
           profiles: {
-            [meta.username]: initGroupProfile(meta.identityContractID)
+            [meta.username]: initGroupProfile(meta.identityContractID, data.joined_ms)
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: use `meta.createdDate`

File: frontend/model/contracts/group.js [hunk 6]
```diff
@@ -436,7 +451,7 @@ DefineContract({
         if (Object.keys(invite.responses).length === invite.quantity) {
           invite.status = INVITE_STATUS.USED
         }
-        Vue.set(state.profiles, meta.username, initGroupProfile(meta.identityContractID))
+        Vue.set(state.profiles, meta.username, initGroupProfile(meta.identityContractID, data.joined_ms))
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: use `meta.createdDate`

File: frontend/views/components/Avatar.vue [hunk 7]
```diff
@@ -1,11 +1,13 @@
 <template lang='pug'>
   img.c-avatar(
+    v-if='imageURL'
```
Comment (taoeffect) [inline frontend/views/components/Avatar.vue:3]: Why is needed? Seemed to be working fine before?

File: frontend/views/components/AvatarUser.vue [hunk 8]
```diff
@@ -38,7 +38,7 @@ export default {
       console.debug(`Looking for ${this.username} profile picture`)
       const userContractId = await sbp('namespace/lookup', this.username)
       if (!userContractId) {
-        console.error(`AvatarUser: ${this.username} doesn't exist!`)
+        console.warn(`AvatarUser: ${this.username} doesn't exist!`)
```
Comment (taoeffect) [inline frontend/views/components/AvatarUser.vue:41]: Why the change?

File: frontend/views/containers/dashboard/GroupMembers.vue [hunk 9]
```diff
@@ -54,45 +42,69 @@
...
-      'ourUsername'
+      'ourUsername',
+      'userDisplayName',
+      'currentGroupState',
+      'groupMembersPending'
     ]),
+    dateNow () {
+      console.log('calculate dateNow')
+      return Date.now()
+    },
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembers.vue]: no reason for creating `dateNow`, please remove

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue [hunk 10]
```diff
@@ -89,17 +96,31 @@ export default {
...
       'globalProfile',
-      'groupMembersCount'
+      'groupMembersCount',
+      'groupMembersPending',
+      'ourUsername'
     ]),
+    dateNow () {
+      console.log('calculate dateNow')
+      return Date.now()
+    },
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembersAllModal.vue]: no reason for creating `dateNow`, please remove

File: frontend/views/containers/dashboard/GroupMembersTooltipPending.vue [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<template lang="pug">
+  tooltip(direction='bottom-end')
+    span.button.is-icon-small(data-test='pendingTooltip')
+      i.icon-question-circle
+    template(slot='tooltip')
+      i18n(
+        tag='p'
+        :args='{ invitedBy: `${invitedBy} ${invitedBy === ourUsername ? `(${L("you")})` : ""}` }'
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembersTooltipPending.vue]: Simply:

```
:args='{ username }'
```

This sentence will not translate into all language properly otherwise.

Alternatively, you can construct two full sentences:

1. `This member did not use their invite link to join the group yet. This link should be given to them by {username}`
2. `This member did not use their invite link to join the group yet. This link should be given to them by {username} (you)`

By using computed properties and the `L` function

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue [hunk 12]
```diff
@@ -113,6 +134,15 @@ export default {
...
   },
   methods: {
+    isNewMember (username) {
+      if (username === this.ourUsername) {
+        return false
+      }
+      const weJoined = this.currentGroupState.profiles[this.ourUsername].joined_ms
+      const memberJoined = this.currentGroupState.profiles[username].joined_ms
+      const joinedAfterUs = weJoined < memberJoined
+      return joinedAfterUs && this.dateNow - memberJoined < 604800000 // joined less than 1w (168h) ago.
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembersAllModal.vue]: You can use `moment` on the ISOString

File: frontend/model/contracts/group.js [hunk 13]
```diff
@@ -149,6 +150,18 @@ DefineContract({
     groupMembersCount (state, getters) {
       return getters.groupMembersByUsername.length
     },
+    groupMembersPending (state, getters) {
+      const invites = getters.currentGroupState.invites
+      const invitesValid = Object.keys(invites).filter(id => invites[id].status === INVITE_STATUS.VALID && invites[id].creator !== INVITE_INITIAL_CREATOR)
+      const pendingMembers = {}
+      for (const secretId of invitesValid) {
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: This is currently doing 3 loops:

1. `Object.keys(invites)`
2. `.filter(...)`
3. `for (const secretId of ...`

Replace 3 loops with a single loop:

```js
for (const inviteId in invites) {
  let invite = invites[inviteId]
  if (invite.status === INVITE_STATUS.VALID && invite.creator !== INVITE_INITIAL_CREATOR) {
    // stuff below
  }
}
```

File: frontend/model/contracts/group.js [hunk 14]
```diff
@@ -149,6 +150,18 @@ DefineContract({
...
+        pendingMembers[invites[secretId].invitee] = {
+          isPending: true,
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: No reason to add `isPending: true`, remove. Replace with maybe the `inviteId`.

File: frontend/model/contracts/group.js [hunk 15]
```diff
@@ -173,6 +186,7 @@ DefineContract({
     'gi.contracts/group': {
       validate: objectMaybeOf({
         invites: mapOf(string, inviteType),
+        joined_ms: number,
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: remove (rely on existing `meta.createdDate`) (avoid creating new variables if they are not needed)

**Review by taoeffect — CHANGES_REQUESTED** [review]
First draft review done!

File: frontend/views/components/Avatar.vue (hunk 7 above)
Comment (sandrina-p) [reply frontend/views/components/Avatar.vue:3]: note the `v-else`. an image must always have a `src`, otherwise it's invalid. so, when it doesn't have a src, it only renders a placeholder (a div)

File: frontend/views/components/AvatarUser.vue (hunk 8 above)
Comment (sandrina-p) [reply frontend/views/components/AvatarUser.vue:41]: Because I think this is not an error... Nothing wrong happened with the app... It's just an edge case. This can happen if we try to look for a user that doesn't exist.
Example1: looking for a member name that was invited but it didn't join the group yet.) Example2: looking for a username that doesn't exist anymore (deleted the account)

File: frontend/views/containers/dashboard/GroupMembersTooltipPending.vue (hunk 11 above)
Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersTooltipPending.vue]: Using the L function? How? I came up with this...
```pug
      i18n(
        ref='msgYou'
        v-if='invitedBy === ourUsername'
        tag='p'
        :args='{ invitedBy }'
      ) This member did not use their invite link to join the group yet. This link should be given to them by {invitedBy} (you).
      i18n(
        v-else
        ref='msg'
        tag='p'
        :args='{ invitedBy }'
      ) This member did not use their invite link to join the group yet. This link should be given to them by {invitedBy}.
```

Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembersTooltipPending.vue]: That uses way too many lines, so in this case a computed property is better. See for example `introTitle` in Payments.vue.

File: frontend/model/contracts/group.js (hunk 15 above)
Comment (sandrina-p) [reply frontend/model/contracts/group.js]: I didn't remember we had `meta.createdDate`, will change!

Comment (sandrina-p) [reply frontend/model/contracts/group.js]: I didn't remember we had meta.createdDate, will change!

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue (hunk 12 above)
Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: I'm not familiar with `moment`. What method do you think it's the best for this case?

File: frontend/views/containers/dashboard/GroupMembersTooltipPending.vue (hunk 11 above)
Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersTooltipPending.vue]: That was what I thought. doing it

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue (hunk 12 above)
Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: - https://momentjs.com/docs/#/manipulating/
- https://momentjs.com/docs/#/parsing/date/
- https://momentjs.com/docs/#/parsing/string/

You can `import moment from 'moment'` at the top (it seems to come from one of our dependencies... not sure how)

Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: Hum... it seems an elephant for what we need here. But found a way more or less intuitive:

```js
moment(memberJoined).isBefore(this.now, 'week')
```

this.now is a computed property returning `moment()`. It should be a computed property to avoid "calculating" now N times in a N loop. Specially on the list with all members (and search).

Btw, I looked into our source and `moment` is an indirect development dependency (it means we don't import it on package.json). The package importing it is Cypress!

So... if we really want to use `moment` it should be part of our dependencies. Just a reminder that this package is [65kb minified](https://bundlephobia.com/result?p=moment@2.24.0). I believe it can be useful to display a more "human time"  on proposals and chat... but it's way heavier than I thought.

Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: I agree that `moment` is probably something we can live without. If you would like to switch back to using the microseconds approach feel free, just initialize the date object using the string (`new Date(timestamp).getTime()`) and then you can use it with `Date.now()`.

I still do not think a computed property is needed. The caching isn't needed in this case I do not think.

Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: I'm okay with leaving it with `moment`. For now, it's more than okay. I just wanted to show my concern...

> I still do not think a computed property is needed. The caching isn't needed in this case I do not think.

Well, it's the difference between calculating "Date.now()" n times vs 1 time. When n times is the number of members and every time we do a search. I think some cache doesn't hurt.

Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: > Well, it's the difference between calculating "Date.now()" n times vs 1 time

That's not an issue since `Date.now()` doesn't take any significant resources.  Also, we want to make sure that it's updated eventually and isn't cached forever, which is what might happen with `this.now` since that doesn't depend on any Vue.js variables, and so it might return the same value forever.

Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: If you can implement this without `moment` it would save us time in the future too, since we want to get removing unnecessary dependencies the closer we get to releasing this.

Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue]: > Also, we want to make sure that it's updated eventually and isn't cached forever

That's not how component's computed properties work. They are "cached" only while the component is mounted. Once it's destroyed (ex: the page changes or the modal is closed), it will re-calculate everything again.

But anyway... changed back to "old school `Date.now()`" and without using a computed property. We have other performance issues more important than this micro-optimization ;)

File: frontend/model/contracts/group.js [hunk 16]
```diff
@@ -62,10 +62,11 @@ function initMonthlyPayments () {
   }
 }
 
-function initGroupProfile (contractID: string) {
+function initGroupProfile (contractID: string, joined: ?number) {
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: `joined: string`

File: frontend/model/contracts/group.js [hunk 17]
```diff
@@ -62,10 +62,11 @@ function initMonthlyPayments () {
...
   return {
     contractID: contractID,
-    nonMonetaryContributions: []
+    nonMonetaryContributions: [],
+    joined_ms: joined
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: `joinedDate`

File: frontend/views/containers/dashboard/GroupMembers.vue [hunk 18]
```diff
@@ -54,45 +42,64 @@
...
+    invite () {
+      this.$router.push({ path: '/invite' })
+    },
+    openModal (modal, queries) {
+      sbp('okTurtles.events/emit', OPEN_MODAL, modal, queries)
+    },
+    isNewMember (username) {
+      if (username === this.ourUsername) { return false }
+
+      const memberJoined = this.currentGroupState.profiles[username].joined_ms
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembers.vue]: This needs to be converted to milliseconds

```js
const joined_ms = new Date(this.currentGroupState.profiles[username].joinedDate).getTime()
```

**Review by taoeffect — CHANGES_REQUESTED** [review]
quick feedback re date usage

File: frontend/views/containers/dashboard/GroupMembers.vue (hunk 18 above)
Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembers.vue]: Changed where needed

File: package.json [hunk 19]
```diff
@@ -101,6 +101,7 @@
     "buffer": "5.4.3",
     "form-data": "2.5.1",
     "localforage": "1.7.3",
+    "moment": "2.24.0",
```
Comment (taoeffect) [inline package.json]: `moment` is something we probably won't end up using since it's such a big dependency, so plz remove & reset `package-lock.json`

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue [hunk 20]
```diff
@@ -89,17 +96,30 @@ export default {
...
-      'groupMembersCount'
+      'groupMembersCount',
+      'groupMembersPending',
+      'ourUsername'
     ]),
+    weJoinedMs () {
+      return new Date(this.currentGroupState.profiles[this.ourUsername].joinedDate).getTime()
+    },
+    allMembers () {
+      return Object.keys({ ...this.groupMembersPending, ...this.groupProfiles })
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembersAllModal.vue:110]: If you just want the keys of these two objects, then there's a faster way of doing this:

```js
return Object.keys(this.groupMembersPending).concat(Object.keys(this.groupProfiles))
```

File: package.json (hunk 19 above)
Comment (sandrina-p) [reply package.json]: I added `moment` to our dependencies because it's something we are actually using at MonthOverview, PaymentDetails and PaymentsList components. And all direct dependencies must be linked on `package.json`

If we really want to stop using it (removing it), that's an issue for another PR, not this one. :)

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue (hunk 20 above)
Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue:110]: Hum... that's what I'm doing here. 😅 The difference to your suggestion is that I'm using destructuring (`...`) to create the array instead of using `Object.keys` twice and `.concat`.

And using destructuring is faster: https://jsperf.com/object-keys-concact-vs-destructuring/1

File: package.json (hunk 19 above)
Comment (taoeffect) [reply package.json]: It was working fine before this PR and it will work fine after you remove it here... please remove. We want to get rid of as many dependencies as humanely possible.

Comment (taoeffect) [reply package.json]: That other issue exists in the form of #372. However I've created a new issue specifically for moment (#874). Please still remove it here tho.

File: frontend/views/containers/dashboard/GroupMembersAllModal.vue (hunk 20 above)
Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembersAllModal.vue:110]: > And using destructuring is faster: https://jsperf.com/object-keys-concact-vs-destructuring/1

Huh. That's surprising. OK we'll keep it your way 😄  👍

File: package.json (hunk 19 above)
Comment (sandrina-p) [reply package.json]: > It was working fine before this PR and it will work fine after you remove it here

It works fine because we have Cypress. If tomorrow Cypress doesn't need `moment` anymore, GI installation will break. Btw, there is an existing JS rule on eslint and node community called "no-extraneous-dependencies". See [ref 1](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md) and [ref 2](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-missing-import.md)

But anyway... removed.

File: frontend/model/contracts/group.js [hunk 21]
```diff
@@ -423,7 +437,7 @@ DefineContract({
     },
     'gi.contracts/group/inviteAccept': {
       validate: objectOf({
-        inviteSecret: string // NOTE: simulate the OP_KEY_* stuff for now
+        inviteSecret: string // NOTE: simulate the OP_KEY_* stuff for now,
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: random comma?

File: frontend/model/contracts/group.js (hunk 16 above)
Comment (sandrina-p) [reply frontend/model/contracts/group.js]: done

File: frontend/views/containers/dashboard/GroupMembers.vue [hunk 22]
```diff
@@ -13,34 +13,22 @@
...
-          i.icon-question-circle
-        template(slot='tooltip')
-          i18n(
-            tag='p'
-            :args='{ username }'
-          ) We are waiting for {username} to join the group by using their unique invite link.
-
-      group-member-menu(v-else :username='username')
+        | {{ userDisplayName(username) }}&nbsp;
+        i18n(v-if='username === ourUsername') (you)
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembers.vue]: To guard against weirdness in other languages (like the positioning of `(you)`), can use a method instead?

E.g.

```pug
.c-name.has-ellipsis(data-test='username') {{ localizedName(username) }}
```

Then:

```js
methods: {
  localizedName (username) {
    const name = this.userDisplayName(username)
    return username === this.ourUsername ? L('{name} (you)', { name }) : name
  }
}
```

File: frontend/views/containers/dashboard/GroupMembers.vue [hunk 23]
```diff
@@ -54,45 +42,64 @@
...
     ]),
+    weJoinedMs () {
+      return new Date(this.currentGroupState.profiles[this.ourUsername].joinedDate).getTime()
+    },
     firstTenMembers () {
       const profiles = this.groupProfiles
-      const usernames = Object.keys(profiles).slice(0, 10)
-      return usernames.reduce((acc, username) => {
+      const sliceIndex = 10 - Math.min(10, Object.keys(this.groupMembersPending).length) // avoid slicing too many members.
+      const usernames = Object.keys(profiles).slice(0, sliceIndex)
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembers.vue:73]: Is this change necessary? AFAICT `.slice` will automatically bounds check for you, you don't need to do it yourself.

File: frontend/views/containers/payments/PaymentsList.vue [hunk 24]
```diff
@@ -29,7 +29,7 @@ table.table.table-in-card.c-payments(:class='{"is-editing": paymentsType === "ed
 
         // TODO: replace condition to indicate whether or not the payment date is < or > than the current date using payment.paymentStatusText
         i18n.c-user-month(
-          :class='index === 0 ? "has-text-1" : "pill is-danger is-small"'
+          :class='index === 0 ? "has-text-1" : "pill is-danger"'
```
Comment (taoeffect) [inline frontend/views/containers/payments/PaymentsList.vue:32]: Are the changes in this file (e.g. removal of `is-small`) because the previous version didn't follow the designs?

File: frontend/model/contracts/group.js (hunk 21 above)
Comment (taoeffect) [inline frontend/model/contracts/group.js]: Remove random comma?

**Review by taoeffect — CHANGES_REQUESTED** [review]
Tested, works great. 💪 

Only minor changes requested.

File: frontend/views/containers/dashboard/GroupMembers.vue (hunk 23 above)
Comment (sandrina-p) [reply frontend/views/containers/dashboard/GroupMembers.vue:73]: No... Let's imagine we have an array `[1,2,3,4,5]` and we do `slice(0, -3)`, the output will be `[1, 2]`. IRL, this can happen if we have more than 10 members pending. Run the code on your computer and see it by yourself.

File: frontend/views/containers/payments/PaymentsList.vue (hunk 24 above)
Comment (sandrina-p) [reply frontend/views/containers/payments/PaymentsList.vue:32]: `is-small` isn't needed. It doesn't exist a "small" variant of the `pill` component. They are all the same size.

File: frontend/views/containers/dashboard/GroupMembers.vue (hunk 23 above)
Comment (taoeffect) [reply frontend/views/containers/dashboard/GroupMembers.vue:73]: Oh, I see what you mean. Oops, misread the code. Nice job! 👍

**Review by taoeffect — APPROVED** [review]
Great job @sandrina-p!
